openflighthpc/flight-profile
# Conditional Dpendencies Support

# Overview

Previously, the dependencies are defined as a fixed array through each identity's metadata. However, some types of clusters require setting the dependencies dynamically based on the user's profile configuration. This PR introduces the `dependencies` field to the cluster type question's metadata to support such requirements.

# Major Changes

- `Configure#update_dependencies` is added to the `profile configure` command process.
- `Type#set_conditional_dependencies` is added to connect `Configure` and `Identity` classes.
- `Identity#set_conditional_dependencies` is added to write the conditional dependency into the file.
- `Identity#all` is modified to include the conditional dependencies.

# Testing

1. Run `vim /opt/flight/usr/lib/profile/types/<specific-type>/metadata.yaml` to modify a random question of a specific cluster type:
   ```
   # The current question should look like this
   id: cluster_name
   env: CLUSTERNAME
   text: 'Cluster name:'
   default: my-cluster
   validation:
     type: string
   dependencies:    # append this section to the question
     - identity: login   # assume this cluster type have identities 'login' and 'ipa'
       depend_on:
          - ipa
       where: flight
   ```
2. Run `flight profile configure` and answer the question `cluster_name` with the answer `flight`.
3. Run `flight profile identities`. The identity `ipa` should be listed as one of the dependencies of `login`.
4. Run `flight profile configure` and _change_ the answer to the question `cluster_name`.
5. Run `flight profile identities`. The identity `ipa` should _NO LONGER_ be listed as one of the dependencies of `login`.
